Add validation for uploaded and synced manifests

lubosmj opened the pull request:
In this commit, a couple of validation schemas were
introduced to the sync and push workflows. Newly added
manifests or manifest lists are now being validated by
the JSON validator.

closes #854
closes #853
closes #672

lubosmj commented:
The reported problems with invalid manifest types are now gone (https://github.com/pulp/pulp_container/issues/854, https://github.com/pulp/pulp_container/issues/853):
```
(pulp) [vagrant@pulp3-source-fedora35 ~]$ podman login localhost:24817 --tls-verify=false
Username: admin
Password: 
Login Succeeded!
(pulp) [vagrant@pulp3-source-fedora35 ~]$ podman manifest push --all   report-bug localhost:24817/foo:report-bug --format oci --tls-verify=false
Getting image list signatures
Copying 0 of 0 images in list
Writing manifest list to image destination
Error: Uploading manifest list failed, attempted the following formats: application/vnd.oci.image.index.v1+json(uploading manifest report-bug to localhost:24817/foo: manifest invalid: manifest invalid)
(pulp) [vagrant@pulp3-source-fedora35 ~]$ podman -v
podman version 3.4.7
(pulp) [vagrant@pulp3-source-fedora35 ~]$ podman manifest push --all   report-bug localhost:24817/foo:report-bug --format v2s2 --tls-verify=false
Getting image list signatures
Copying 0 of 0 images in list
Writing manifest list to image destination
Storing list signatures
(pulp) [vagrant@pulp3-source-fedora35 ~]$ podman manifest push --all   report-bug localhost:24817/foo:report-bug --format oci --tls-verify=false
Getting image list signatures
Copying 0 of 0 images in list
Writing manifest list to image destination
Error: Uploading manifest list failed, attempted the following formats: application/vnd.oci.image.index.v1+json(uploading manifest report-bug to localhost:24817/foo: manifest invalid: manifest invalid)
```

lubosmj commented on pulp_container/app/json_schemas.py:
```diff
@@ -1,54 +1,210 @@
+                        "type": "string",
+                        "enum": [
+                            "application/vnd.docker.image.rootfs.diff.tar.gzip",
+                            "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
```
I believe we do not support foreign layers. Does it mean that I should remove this entry from the enum, @ipanova?

ipanova replied:
we do not download during sync them but in manifest schema they remain since  we do not change the manifest

ipanova reviewed (commented):
This is looking great overall, thank you for adding the validation!
I don't see validation for schema1 will you add it too?

ipanova commented:
It would be great to add this validation to the sync worklow too. Will you do it as part of this PR or in a separate one?

lubosmj commented:
> It would be great to add this validation to the sync worklow too. Will you do it as part of this PR or in a separate one?

I would love to do that in a separate PR.

I will add validation for the v1 schema.

lubosmj commented:
@ipanova, I have added a schema for v1 manifests (I think they share schema with signed v1 manifests) and added the validation to the sync pipeline. How do you like it now? Do I need to attach a fourth issue for adding validation during the sync procedure?

I also wonder what are we going to do with https://github.com/pulp/pulp_container/issues/648. Because it will not be doable anymore after adding the validation. Also, corrupted manifests may be referenced from other manifests. The question is: do we really need to allow users to perform "partial" sync?

ipanova commented on CHANGES/882.bugfix:
```diff
@@ -0,0 +1,2 @@
+Fixed a bug raised when syncing manifest lists that contain only the reference to manifests but
```
Fixed a bug raised when syncing manifest lists that had emtpy list of manifests

ipanova commented on pulp_container/app/tasks/sync_stages.py:
```diff
@@ -129,10 +134,19 @@ async def run(self):
+                    schema = determine_schema(media_type)
+                except ValueError:
+                    raise RuntimeError(
+                        f"Requested a validation schema for an unknown media type: '{media_type}'"
```
I'd also raise here ManifestInvalid error

ipanova replied:
Can we re-use validate_manifest in both sync and push codepaths? expose it in utils.

ipanova commented on pulp_container/app/tasks/sync_stages.py:
```diff
@@ -315,7 +329,7 @@ async def handle_blobs(self, manifest_dc, content_data):
             await self.put(blob_dc)
 
-    def create_tagged_manifest_list(self, tag_name, saved_artifact, manifest_list_data):
+    def create_tagged_manifest_list(self, tag_name, saved_artifact, manifest_list_data, media_type):
```
what about image_manifest?oci format might not have media_type present too

ipanova commented on pulp_container/app/tasks/sync_stages.py:
```diff
@@ -129,10 +134,19 @@ async def run(self):
                 if media_type in (MEDIA_TYPE.MANIFEST_LIST, MEDIA_TYPE.INDEX_OCI):
                     list_dc = self.create_tagged_manifest_list(
-                        tag_name, saved_artifact, content_data
+                        tag_name, saved_artifact, content_data, media_type
```
you don't seem to validate listed_manifests

ipanova commented on pulp_container/app/registry_api.py:
```diff
@@ -895,6 +892,20 @@ def put(self, request, path, pk=None):
 
         content_data = json.loads(raw_data)
 
+        try:
```
There are also other places where you can use manifest schema validation and get rid of the extra if/else. I'll take the conversation offline.

ipanova replied:
same for Sync codepath

ipanova commented on pulp_container/app/json_schemas.py:
```diff
@@ -1,54 +1,253 @@
+                [
+                    "application/vnd.oci.image.layer.v1.tar",
+                    "application/vnd.oci.image.layer.v1.tar+gzip",
+                    "application/vnd.oci.image.layer.v1.tar+zstd",
```
we need to update the constants too then

ipanova commented:
your PR might fix this issue too https://github.com/pulp/pulp_container/issues/883

ipanova commented on pulp_container/app/tasks/sync_stages.py:
```diff
@@ -129,10 +134,19 @@ async def run(self):
 
                 content_data = json.loads(raw_data)
-                media_type = content_data.get("mediaType")
+                media_type = content_data.get("mediaType") or dl_res.headers["content-type"]
```
https://testsuite.io/content-type-header
dl_res.headers.get("content-type")

ipanova replied:
@lubosmj can you  extract this fix in a separate PR and mark it as bugfix so we can backport it later https://github.com/pulp/pulp_container/issues/883#issuecomment-1193869565

ipanova commented on pulp_container/app/json_schemas.py:
```diff
@@ -1,54 +1,253 @@
+                    "platform": {
+                        "type": "object",
+                        "properties": {
+                            "architecture": {"type": "string"},
```
maybe not in this PR

ipanova commented on pulp_container/app/json_schemas.py:
```diff
@@ -1,54 +1,253 @@
+                    "application/vnd.oci.image.manifest.v1+json",
+                    "application/vnd.oci.image.index.v1+json",
+                    # there might be a case where an index contains manifests of type docker.dist
+                    "application/vnd.docker.distribution.manifest.v2+json",
```
We will follow OCI specs

ipanova commented:
> @ipanova, I have added a schema for v1 manifests (I think they share the schema with signed v1 manifests) and added the validation to the sync pipeline. How do you like it now? Do I need to attach a fourth issue for adding the validation during the sync procedure?

Re-using this isseure should be fine to include validation for synced and uploaded content.
> 
> **Unrelated:** I also wonder what are we going to do with #648. Because it will not be easily doable anymore after adding the validation (e.g., initiating a skip task for a parent manifest list that was already parsed and validated). The question is: do we really need to allow users to perform "partial" sync?
That's a good question, we will need to evaluate the complexity of the #648 but from the user perspective, this would be a handy sync behavior.

ipanova commented on pulp_container/app/json_schemas.py:
```diff
@@ -1,54 +1,254 @@
+                    MEDIA_TYPE.REGULAR_BLOB_OCI_TAR_GZIP,
+                    MEDIA_TYPE.REGULAR_BLOB_OCI_TAR_ZSTD,
+                    MEDIA_TYPE.FOREIGN_BLOB_OCI_TAR,
+                    MEDIA_TYPE.FOREIGN_BLOB_OCI_TAR_GZIP,
```
you forgot foreign_blob zstd compression

ipanova commented on pulp_container/app/utils.py:
```diff
@@ -156,3 +164,36 @@ def determine_media_type(content_data, response):
+        schema_validator = determine_schema(content_type)
+    except ValueError:
+        raise ManifestInvalid(
+            reason=f"A manifest of an unknown content type was provided: {content_type}",
```
unknown media type

ipanova commented on pulp_container/app/utils.py:
```diff
@@ -90,8 +98,8 @@ def extract_data_from_signature(signature_raw, man_digest):
-    for error in validator.iter_errors(sig_json):
-        errors.append(error.message)
+    for error in signature_validator.iter_errors(sig_json):
+        errors.append(f'{".".join(error.path)}: {error.message}')
```
what's contained in error.path?

lubosmj replied:
The name of the JSON field which is considered invalid (e.g., the error message would look like this in case of an invalid schema version: "schemaVersion: 1 is less than 2").

ipanova replied:
aha, gotcha, thanks!

ipanova commented on pulp_container/constants.py:
```diff
@@ -12,8 +12,11 @@
-    REGULAR_BLOB_OCI="application/vnd.oci.image.layer.v1.tar+gzip",
-    FOREIGN_BLOB_OCI="application/vnd.oci.image.layer.nondistributable.v1.tar+gzip",
+    REGULAR_BLOB_OCI_TAR="application/vnd.oci.image.layer.v1.tar",
+    REGULAR_BLOB_OCI_TAR_GZIP="application/vnd.oci.image.layer.v1.tar+gzip",
```
:+1:

ipanova commented on pulp_container/app/tasks/sync_stages.py:
```diff
@@ -587,7 +594,7 @@ def _include_layer(self, layer):
         foreign_excluded = not self.remote.include_foreign_layers
         layer_type = layer.get("mediaType", MEDIA_TYPE.REGULAR_BLOB)
-        is_foreign = layer_type in (MEDIA_TYPE.FOREIGN_BLOB, MEDIA_TYPE.FOREIGN_BLOB_OCI)
+        is_foreign = layer_type in (MEDIA_TYPE.FOREIGN_BLOB, MEDIA_TYPE.FOREIGN_BLOB_OCI_TAR_GZIP)
```
i am afraid you will need to include here all the types of compression

lubosmj replied:
I appended only `MEDIA_TYPE.FOREIGN_BLOB_OCI_TAR_ZSTD`.

ipanova commented on pulp_container/app/registry_api.py:
```diff
@@ -869,34 +873,27 @@ def put(self, request, path, pk=None):
 
         content_data = json.loads(raw_data)
 
+        validate_manifest(content_data, request.content_type, manifest_digest)
```
I know that were relying on request.content_type before, but using here determine_media_type won't hurt, wdyt?

ipanova commented on pulp_container/app/registry_api.py:
```diff
@@ -935,14 +932,13 @@ def put(self, request, path, pk=None):
+            if not config_layer:
+                raise ManifestInvalid(
+                    digest=manifest_digest,
+                    reason="Pushing manifests of the version V1 is not supported",
```
:+1:

ipanova commented on pulp_container/app/registry_api.py:
```diff
@@ -954,23 +950,7 @@ def put(self, request, path, pk=None):
-                media_type = layer.get("mediaType")
-                urls = layer.get("urls")
                 digest = layer.get("digest")
-                if (
```
i know this is not explicit, but we need to keep this check https://github.com/quay/quay/blob/ef91c57c233e9d23372afac34c0d062a38171ffa/image/docker/schema2/manifest.py#L165
To be honest i don't understand why `urls` are not required when blob is foreign in the schema validation
maybe here you can add `urls` as additional_required if media_type == foreign_blob_types?

lubosmj replied:
Actually, this cannot be done dynamically (as you suggested: if media_type == foreign_blob_types, because we need to initialize this beforehand). I have seen conditional `required` only for keys, but not values. I am adding back the if condition.

ipanova replied:
i see :/

ipanova commented on pulp_container/app/registry_api.py:
```diff
@@ -869,34 +874,28 @@ def put(self, request, path, pk=None):
+        # when a user uploads a manifest list with zero listed manifests (no blobs were uploaded
+        # before) and the specified repository has not been created yet, create the repository
+        # without raising an error
+        create_new_repo = request.content_type in (
```
s/request.content_type/media_type

ipanova commented on pulp_container/app/registry_api.py:
```diff
@@ -869,34 +874,28 @@ def put(self, request, path, pk=None):
+        )
+        _, repository = self.get_dr_push(request, path, create=create_new_repo)
+
         if request.content_type in (
```
s/request.content_type/media_type

ipanova replied:
https://github.com/pulp/pulp_container/pull/866/files#diff-1b944e15118f82233aa03a435e9e05c188ad0d371b71be43260506b4a5eeddffL914

ipanova replied:
https://github.com/pulp/pulp_container/pull/866/files#diff-1b944e15118f82233aa03a435e9e05c188ad0d371b71be43260506b4a5eeddffR982

ipanova reviewed (approved):
Thank you! This work will greatly improve the quality of the sync and upload pipelines! :fireworks:

ipanova commented:
To make it a meaningful number of comments to 45 ;) ( only by @lubosmj  request) I will leave here some more :monkey:
